docs/public_ips.md

```diff
@@ -0,0 +1,289 @@
+setting is on the kube-proxy machines (the minions), so we cannot prevent the user setting an overlapping range.
+The salt scripts could check this.  If an overlapping range is set, conflicts will randomly happen between
+cluster services and public services on some kube-proxy instances, and those kube-proxies will log errors and
+fail to serve that service.  This will likely result in degraded service availability.
+
+## API evolution
+
+The main changes:
+
+* we deprecate createExternalLoadBalancer, and replace it with visibility
+* we move publicIPs to the Status
+* we add a field hostPort to each service port
+ 
+Mappings:
+
+`visibility` is mappable:
+
+```
+new.visibility = old.createExternalLoadBalancer ? "loadbalancer" : "cluster"
+```
+
+```
+old.createExternalLoadBalancer = (new.visibility == 'loadbalancer')
+```
+
+`publicIPs` will be copied to the Spec if set on the Status, and from the Spec to the Status if not set on the Status.
+Once each cloud provider sets the Status, we will stop copying from the Spec to the Status.
+
+For v1 final:
+* we should remove publicIPs from the Spec, and require that this information is retrieved through the Status.
+However, we would need to change the openstack code to auto-allocate a VIP.
+* we should remove createExternalLoadBalancer.
+
+## kube-proxy pseudocode
+
+
+```
+for each service {
+    for each service.port {
+      // listen on cluster
+      listen on <internalMagicIp>:0
+      proxyPort = figureOutTheRandomPortWeJustGotAssigned()
+      openPortal(service.portalIP, port.port, proxyPort)
+
+      for each service.publicIP {  // deprecated
+        openPublicPortal(publicIP, port.port, proxyPort) // uses cloud-provider to decide srcip vs dstip (gclb vs rax)
+      }
+
+      if service.spec.visibility == "loadbalancer" {
+        if (port.hostPort != 0) {
+          openLoadBalancerPortals(service, port.port, port.hostPort) // uses cloud-provider to figure out correct iptables
+        }
+      } else if service.spec.visibility == "public" {
+        if (port.hostPort != 0) {
+          openPublicPortal(<publicMagicIp>, port.port, port.hostPort)
+        }
+      }
+   }
+}
+```
+
+
+## kube-proxy iptables
+
+The current implementation of kube-proxy allocates ports for cluster services, randomly from the range
+in /proc/sys/net/ipv4/ip_local_port_range.  By allocating ports from a non-overlapping range specified by
+the public_service_ports argument, we should avoid conflicts.
+
+(There may be changes we can make to avoid this in future, but they are orthogonal to the public-ports work)
+
+It is possible that there would be an unexpected port-conflict meaning we could not open the port.  This will be treated
+as an unexpected failure (like out-of-fds); it will be logged and kube-proxy will continue and likely retry.  This should
+only happen if we run new services on the minions or specify public_service_ports overlapping with ip_local_port_range.
+
+
+## Sidecar-pod implementations
```

Comment:
We usually use "sidecar" to refer to containers running in users' pods. Is that what you mean?